File: android/src/main/java/com/jimmydaddy/imagemarker/base/Position.java
package com.jimmydaddy.imagemarker.base;

/**
 * Created by jimmydaddy on 2017/9/23.
 */

public class Position {
  public float X;
  public float Y;

  public Position(float x, float y) {
    X = x;
    Y = y;
  }

  public float getX() {
    return X;
  }

  public void setX(float x) {
    X = x;
  }

  public float getY() {
    return Y;
  }

  public void setY(float y) {
    Y = y;
  }

  static public Position getTextPosition(String position, int margin, int width, int height, int textWidth, int textHeight) {
    if (position == null) {
      return new Position(margin, margin);
    }
    switch (position) {
      case "topCenter":
        return new Position((width - textWidth) / 2, margin);
      case "topRight":
        return new Position(width - textWidth, margin);
      case "center":
        return new Position((width - textWidth) / 2, (height - textHeight) / 2);
      case "bottomLeft":
        return new Position(0, height - textHeight - margin);
      case "bottomCenter":
        return new Position((width - textWidth) / 2, height - textHeight);
      default:
        return new Position(width - textWidth - margin, height - textHeight - margin);
    }
  }

  static public Position getTextPosition(PositionEnum position, int width, int height, int textWidth, int textHeight) {
    // default margin
    int margin = 20;
    if (position == null) {
      return new Position(margin, margin);
    }
    switch (position) {
      case TOP_CENTER:
        return new Position((width - textWidth) / 2, margin);
      case TOP_RIGHT:
        return new Position(width - textWidth, margin);
      case CENTER:
        return new Position((width - textWidth) / 2, (height - textHeight) / 2);
      case BOTTOM_LEFT:
        return new Position(0, height - textHeight - margin);
      case BOTTOM_CENTER:
        return new Position((width - textWidth) / 2, height - textHeight);
      default:
        return new Position(width - textWidth - margin, height - textHeight - margin);
    }
  }

  static public Position getImageRectFromPosition(String position, int width, int height, int imageWidth, int imageHeigt) {
    int left = 20;
    int top = 40;
    int right = imageWidth - width;
    Position pos = new Position(left, top);
    if (position == null) {
      return pos;
    }
    switch (position) {
      case "topCenter":
        left = (imageWidth) / 2 - width / 2;
        pos.setX(left);
        break;
      case "topRight":
        pos.setX(right - 20);
        break;
      case "center":
        left = (imageWidth) / 2 - width / 2;
        top = (imageHeigt) / 2 - height / 2;
        pos.setX(left);
        pos.setY(top);
        break;
      case "bottomLeft":
        top = imageHeigt - height;
        pos.setY(top - 20);
        break;
      case "bottomRight":
        top = imageHeigt - height;
        left = imageWidth - width - 20;
        pos.setX(left - 20);
        pos.setY(top - 20);
        break;
      case "bottomCenter":
        top = imageHeigt - height;
        left = (imageWidth) / 2 - width / 2;
        pos.setX(left - 20);
        pos.setY(top - 20);
    }
    return pos;
  }

  static public Position getImageRectFromPosition(PositionEnum position, int width, int height, int imageWidth, int imageHeigt) {
    int left = 20;
    int top = 40;
    int right = imageWidth - width;
    Position pos = new Position(left, top);
    if (position == null) {
      return pos;
    }
    switch (position) {
      case TOP_CENTER:
        left = (imageWidth) / 2 - width / 2;
        pos.setX(left);
        break;
      case TOP_RIGHT:
        pos.setX(right - 20);
        break;
      case CENTER:
        left = (imageWidth) / 2 - width / 2;
        top = (imageHeigt) / 2 - height / 2;
        pos.setX(left);
        pos.setY(top);
        break;
      case BOTTOM_LEFT:
        top = imageHeigt - height;
        pos.setY(top - 20);
        break;
      case BOTTOM_RIGHT:
        top = imageHeigt - height;
        left = imageWidth - width - 20;
        pos.setX(left - 20);
        pos.setY(top - 20);
        break;
      case BOTTOM_CENTER:
        top = imageHeigt - height;
        left = (imageWidth) / 2 - width / 2;
        pos.setX(left - 20);
        pos.setY(top - 20);
    }
    return pos;
  }

}

File: android/src/main/java/com/jimmydaddy/imagemarker/base/Position.kt
package com.jimmydaddy.imagemarker.base

/**
 * A resolved top-left coordinate in the unrotated background-image coordinate space.
 */
data class Position(var x: Float, var y: Float) {

  companion object {
    private const val DEFAULT_EDGE_INSET = 20f

    fun getTextPosition(
      position: PositionEnum?,
      offsetX: String?,
      offsetY: String?,
      width: Int,
      height: Int,
      textWidth: Int,
      textHeight: Int,
      edgeInset: String? = null
    ): Position = resolve(
      position,
      offsetX,
      offsetY,
      edgeInset,
      width,
      height,
      textWidth,
      textHeight,
      DEFAULT_EDGE_INSET
    )

    @JvmStatic
    fun getImageRectFromPosition(
      position: PositionEnum?,
      offsetX: String?,
      offsetY: String?,
      width: Int,
      height: Int,
      imageWidth: Int,
      imageHeight: Int,
      edgeInset: String? = null
    ): Position = resolve(
      position,
      offsetX,
      offsetY,
      edgeInset,
      imageWidth,
      imageHeight,
      width,
      height,
      0f
    )

    private fun resolve(
      position: PositionEnum?,
      offsetX: String?,
      offsetY: String?,
      edgeInset: String?,
      canvasWidth: Int,
      canvasHeight: Int,
      itemWidth: Int,
      itemHeight: Int,
      unanchoredDefaultInset: Float
    ): Position {
      val explicitX = offsetX?.let { Utils.parseSpreadValue(it, canvasWidth.toFloat()) }
      val explicitY = offsetY?.let { Utils.parseSpreadValue(it, canvasHeight.toFloat()) }

      val configuredInsetX = edgeInset
        ?.let { Utils.parseSpreadValue(it, canvasWidth.toFloat()) }
        ?.coerceAtLeast(0f)
      val configuredInsetY = edgeInset
        ?.let { Utils.parseSpreadValue(it, canvasHeight.toFloat()) }
        ?.coerceAtLeast(0f)

      // Preserve each API's historical unanchored placement. Supplying edgeInset makes omitted
      // axes deterministic across platforms without changing existing calls.
      if (position == null) {
        return Position(
          explicitX ?: configuredInsetX ?: unanchoredDefaultInset,
          explicitY ?: configuredInsetY ?: unanchoredDefaultInset
        )
      }

      val insetX = configuredInsetX ?: DEFAULT_EDGE_INSET
      val insetY = configuredInsetY ?: DEFAULT_EDGE_INSET
      val centeredX = centered(canvasWidth, itemWidth)
      val centeredY = centered(canvasHeight, itemHeight)

      val anchored = when (position) {
        PositionEnum.TOP_LEFT -> Position(insetX, insetY)
        PositionEnum.TOP_CENTER -> Position(centeredX, insetY)
        PositionEnum.TOP_RIGHT -> Position(canvasWidth - itemWidth - insetX, insetY)
        PositionEnum.CENTER -> Position(centeredX, centeredY)
        PositionEnum.BOTTOM_LEFT -> Position(insetX, canvasHeight - itemHeight - insetY)
        PositionEnum.BOTTOM_CENTER -> Position(centeredX, canvasHeight - itemHeight - insetY)
        PositionEnum.BOTTOM_RIGHT -> Position(
          canvasWidth - itemWidth - insetX,
          canvasHeight - itemHeight - insetY
        )
      }

      // Explicit X/Y replace edgeInset for that axis while retaining the anchor's direction.
      if (explicitX != null) {
        anchored.x = when (position) {
          PositionEnum.TOP_LEFT, PositionEnum.BOTTOM_LEFT -> explicitX
          PositionEnum.TOP_CENTER, PositionEnum.CENTER, PositionEnum.BOTTOM_CENTER -> centeredX + explicitX
          PositionEnum.TOP_RIGHT, PositionEnum.BOTTOM_RIGHT -> canvasWidth - itemWidth - explicitX
        }
      }
      if (explicitY != null) {
        anchored.y = when (position) {
          PositionEnum.TOP_LEFT, PositionEnum.TOP_CENTER, PositionEnum.TOP_RIGHT -> explicitY
          PositionEnum.CENTER -> centeredY + explicitY
          PositionEnum.BOTTOM_LEFT, PositionEnum.BOTTOM_CENTER, PositionEnum.BOTTOM_RIGHT -> {
            canvasHeight - itemHeight - explicitY
          }
        }
      }
      return anchored
    }

    private fun centered(canvasSize: Int, itemSize: Int): Float {
      return (canvasSize - itemSize) / 2f
    }
  }
}
